fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/indexes/StandardIndexMaintainer.java

```diff
@@ -405,7 +405,7 @@ protected <M extends Message> void checkUniqueness(@Nonnull FDBIndexableRecord<M
                     Tuple existingEntry = unpackKey(getIndexSubspace(), kv);
                     Tuple existingKey = state.index.getEntryPrimaryKey(existingEntry);
                     if (!TupleHelpers.equals(primaryKey, existingKey)) {
-                        if (state.store.isIndexWriteOnly(state.index)) {
+                        if (isWriteOnlyOrUniquePending()) {
```

Comment:
Are these the semantics we want? I think it makes sense for an index in `READABLE_UNIQUE_PENDING` to remove uniqueness violations on record delete, but arguably, it should enforce the uniqueness constraint during record save to avoid making the problem worse. Or, in other words, the user experience would be something like:

1. They build a unique index, but uniqueness violations mean it is now `READABLE_UNIQUE_PENDING`
2. Immediately, any new record saves start failing if they introduce a new uniqueness violation
3. Record deletes allow the user to start cleaning up the uniqueness violations
4. Once that's completed, the index can be marked as `READABLE`
5. Now queries can start relying on the uniqueness constraint

I'm not entirely convinced either way. I think the current behavior (which is effectively that the uniqueness constraint is ignored until the index can be marked as `READABLE`) is also understandable. We should probably document our decision one way or the other.

Reply:
I strongly agree that new duplications should be forbidden when the index's state is READABLE_UNIQUE_PENDING. I'll implement it in another commit.